parser/internal/parservX/parser_test.go

```diff
@@ -130,3 +150,110 @@ func assertParseFailure(t *testing.T, src string, message string, start int, end
 	assert.Equal(t, start, parseErr.Pos().Start)
 	assert.Equal(t, end, parseErr.Pos().End)
 }
+
+// Copy of testhelpers which allows functions to be passed to the parser
+var versionRegex = regexp.MustCompile(`//groq:version=(\S+)`)
+var paramRegex = regexp.MustCompile(`//groq:param:([^=]+)=([^\n]+)`)
+
+type Test struct {
+	Name      string
+	Query     string
+	Params    groq.Params
+	Functions map[ast.FunctionID]*ast.FunctionDefinition
+	Valid     bool
+	Version   *semver.Range
+	FileName  string
+}
+
+func ASTTest(
+	t *testing.T,
+	test *Test,
+	outputDir string,
+	parser func(query string, params groq.Params, functions map[ast.FunctionID]*ast.FunctionDefinition) (ast.Expression, error),
+) {
+	fileName := filepath.Join(outputDir, SnapshotFileName(test))
+
+	require.NoError(t, os.MkdirAll(filepath.Dir(fileName), 0755))
+
+	t.Logf("File: %s", fileName)
+
+	parsed, err := parser(test.Query, test.Params, test.Functions)
+	require.NoError(t, err)
+
+	type Result struct {
+		Query string
+		AST   ast.Expression
+	}
+
+	result := Result{
+		Query: test.Query,
+		AST:   parsed,
+	}
+
+	actual := []byte(litter.Options{
+		FieldFilter: func(f reflect.StructField, _ reflect.Value) bool {
+			return f.Name != "Source" // Skip the unnecessary source fields
+		},
+		HomePackage: "ast",
+		Compact:     true, // Save space
+	}.Sdump(result))
+
+	expected := testhelpers.ReadSnapshot(t, fileName, actual)
+
+	assert.Equal(t,
+		strings.TrimSpace(string(expected)),
+		strings.TrimSpace(string(actual)),
+	)
+}
+
+func WithEachTest(t *testing.T, f func(t *testing.T, test *Test)) {
+	_, filename, _, _ := runtime.Caller(0)
+
+	pattern := filepath.Join(filepath.Dir(filename), "queries/*.groq")
+	files, err := filepath.Glob(pattern)
```

Comment:
I don't think this is running any of the tests, since filename will be relative to the file where the function is defined.

something like:
```suggestion
	_, filename, _, _ := runtime.Caller(0)

	pattern := filepath.Join(filepath.Dir(filename), "..", "..", "..", "internal", "testhelpers", "queries", "*.groq")
	files, err := filepath.Glob(pattern)
```
?

Reply:
omg true, I'll just move the mother query file in this directory. Doesn't make sense to be separate anymore.